src/statement.ts

```diff
@@ -0,0 +1,240 @@
+import {
+  type ValuesType,
+  type Parameters,
+  type BaseType,
+  getParameters,
+  bindValues,
+} from "./helpers/binding.js";
+import {
+  type AutoWaitConfig,
+  type Config,
+  type Signal,
+  normalize,
+} from "./helpers/index.js";
+import {
+  type ExtractedStatement,
+  type Result,
+  extractReadonly,
+  wrapResult,
+} from "./registry/utils.js";
+import { type ValueOf } from "./validator/query.js";
+import {
+  extractColumn,
+  queryAll,
+  queryFirst,
+  queryRaw,
+  exec,
+} from "./lowlevel.js";
+
+export { type ValuesType, type Parameters, type ValueOf, type BaseType };
+
+export class Statement<S = unknown> {
+  private readonly config: Config & Partial<AutoWaitConfig>;
+  private readonly sql: string;
+  private readonly parameters?: Parameters;
+
+  constructor(
+    config: Config & Partial<AutoWaitConfig>,
+    sql: string,
+    parameters?: Parameters
+  ) {
+    this.config = config;
+    this.sql = sql;
+    this.parameters = parameters;
+  }
+
+  /**
+   * Bind a set of values to the parameters of the prepared statement.
+   * We follow the SQLite convention for prepared statements parameter binding.
+   * We support Ordered (?NNNN), Anonymous (?), and Named (@name, :name, $name) parameters.
+   * @param values A variadic list of values to bind. May include base types, and objects.
+   * @returns A bound Statement.
+   */
+  bind<T = S>(...values: ValuesType[]): Statement<T> {
+    const parameters = getParameters(...values);
+    return new Statement(this.config, this.sql, parameters);
+  }
+
+  /**
+   * Resolve a bound statement to a SQL string.
+   * @returns A valid SQL string.
+   */
+  toString(): string {
```

Comment:
This is _not_ part of the D1 spec, we use it a lot, and I don't like it. Once we have real prepared statements, it'll go away!